plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java

```diff
@@ -111,4 +115,43 @@ public List<Location> descendants(String locationId, Location parentLocation) {
         }
         return location;
     }
+
+    public Bundle getPaginatedLocations(HttpServletRequest request) {
+        String identifier = request.getParameter(Constants.IDENTIFIER);
+        String pageSize = request.getParameter(Constants.PAGINATION_PAGE_SIZE);
+        String pageNumber = request.getParameter(Constants.PAGINATION_PAGE_NUMBER);
+        Map<String, String[]> parameters = new HashMap<>(request.getParameterMap());
+
+        int count =
+                pageSize != null
+                        ? Integer.parseInt(pageSize)
+                        : Constants.PAGINATION_DEFAULT_PAGE_SIZE;
+        int page =
+                pageNumber != null
+                        ? Integer.parseInt(pageNumber)
+                        : Constants.PAGINATION_DEFAULT_PAGE_NUMBER;
+
+        int start = Math.max(0, (page - 1)) * count;
+        Location parentLocation = getLocationById(identifier);
+        List<Location> locations = getDescendants(identifier, parentLocation);
+        List<Resource> resourceLocations = new ArrayList<>(locations);
+        int totalEntries = locations.size();
+
+        int end = Math.min(start + count, resourceLocations.size());
+        List<Resource> paginatedResourceLocations = resourceLocations.subList(start, end);
```

Comment:
In future we may want to figure out an alternative approach to avoid having to re fetch the whole hierarchy for every single page (and then get the sublist). For now once we add the caching it will somewhat alleviate this.